File: java/projects/WebServer.java
package projects;
// PROJECT: Raw Java HTTP Server (no frameworks) | javac WebServer.java && java projects.WebServer
// Build a multi-threaded HTTP/1.1 server using only java.net.ServerSocket

import java.io.*;
import java.net.*;

public class WebServer {

    // TODO 1: Create ServerSocket on port 8080
    //   - new ServerSocket(8080)
    //   - Set SO_REUSEADDR to avoid "Address already in use" on restart

    // TODO 2: Accept connections in a loop
    //   - Socket client = serverSocket.accept();
    //   - Handle each connection in a new Thread (or use ExecutorService pool)

    // TODO 3: Parse HTTP request
    //   - Read first line: "GET /path HTTP/1.1"
    //   - Read headers until blank line "\r\n"
    //   - Store in HashMap<String, String>

    // TODO 4: Router — map paths to handlers
    //   - GET / → 200 "Hello from Java HTTP Server"
    //   - GET /time → 200 with current timestamp (LocalDateTime.now())
    //   - GET /health → 200 {"status":"ok"} with Content-Type: application/json
    //   - else → 404 Not Found

    // TODO 5: Send HTTP/1.1 response
    //   - "HTTP/1.1 200 OK\r\n"
    //   - "Content-Type: text/plain\r\n"
    //   - "Content-Length: N\r\n"
    //   - "Connection: close\r\n"
    //   - "\r\n"
    //   - body

    // TODO 6: Thread pool — use Executors.newFixedThreadPool(10) instead of new Thread per request

    // TODO 7: Graceful shutdown — Runtime.getRuntime().addShutdownHook(new Thread(() -> serverSocket.close()))

    public static void main(String[] args) throws Exception {
        System.out.println("TODO: implement HTTP server on port 8080");
    }
}

File: c/Projects/web_server.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <sys/socket.h>
// PROJECT: HTTP/1.1 Web Server | gcc -o web_server web_server.c && ./web_server
//
// TODO 1: TCP listener on port 8080
//   int sock = socket(AF_INET, SOCK_STREAM, 0);
//   setsockopt(sock, SOL_SOCKET, SO_REUSEADDR, &1, sizeof(1));
//   struct sockaddr_in addr = { .sin_family=AF_INET, .sin_port=htons(8080),
//                               .sin_addr.s_addr=INADDR_ANY };
//   bind(sock, (struct sockaddr*)&addr, sizeof(addr));
//   listen(sock, SOMAXCONN);
//   int conn = accept(sock, NULL, NULL);
//
// TODO 2: HTTP request parsing
//   - Read bytes into buffer: recv(conn, buf, sizeof(buf)-1, 0)
//   - Parse request line: sscanf(buf, "%s %s %s", method, path, version)
//   - Find headers after first "\r\n"; body after "\r\n\r\n"
//
// TODO 3: Router — match path → handler
//   if (strcmp(path, "/") == 0)       serve_index(conn);
//   else if (strcmp(path, "/health")) serve_health(conn);
//   else                              serve_404(conn);
//
// TODO 4: Response builder
//   void respond(int conn, int status, const char *body) {
//       char hdr[512];
//       snprintf(hdr, sizeof(hdr),
//           "HTTP/1.1 %d OK\r\nContent-Length: %zu\r\nConnection: close\r\n\r\n",
//           status, strlen(body));
//       send(conn, hdr, strlen(hdr), 0);
//       send(conn, body, strlen(body), 0);
//   }
//
// TODO 5: Concurrent connections using fork() or pthreads
//   // fork: each accept spawns child process to handle connection
//   // pthread: spawn thread per connection

int main(void) {
    printf("Web Server — TODO: implement\n");
    return 0;
}
